### PR TITLE
Issue #607: shutDown propagation in LuceneSail

### LINKED ISSUE #607: SailRepository.shutDown is not propagated or is not finished.
Affected version **2.0.1**.

I have `java.lang.IllegalStateException: Sail has already been initialized. Ensure this Sail is being used via a Repository.` But it seems the error occurs only in certain conditions because it is only after second use:

The error occurs only after second use of the repository.: 
```
28:33.487 [off-pool-0] DEBUG sparql -- Load record: %s ./tests/resources/22-rdf-syntax-ns.ttl
28:33.516 [off-pool-0] DEBUG triple.loader -- 1. Is repository initialized:  false
28:33.517 [off-pool-0] DEBUG triple.loader -- Initialize repository
28:33.927 [off-pool-0] DEBUG triple.loader -- 2. Is repository initialized:  true
28:33.975 [off-pool-0] DEBUG triple.loader -- Using .add method
28:33.984 [off-pool-0] DEBUG triple.loader -- finish...
28:34.005 [off-pool-0] DEBUG triple.loader -- shutdown repository

# after first use

28:34.006 [off-pool-0] DEBUG sparql -- Load record: %s ./tests/resources/rdf-schema.rdf
28:34.022 [off-pool-0] DEBUG triple.loader -- 1. Is repository initialized:  false
28:34.023 [off-pool-0] DEBUG triple.loader -- Initialize repository
28:34.040 [off-pool-0] DEBUG triple.loader -- 2. Is repository initialized:  true
28:34.047 [off-pool-0] DEBUG triple.loader -- Using .add method
28:34.048 [off-pool-0] DEBUG triple.loader -- finish...
28:34.049 [off-pool-0] DEBUG triple.loader -- shutdown repository

# after second use

28:34.049 [off-pool-0] DEBUG sparql -- Load record: %s ./tests/beet.rdf
28:34.049 [off-pool-0] DEBUG triple.loader -- 1. Is repository initialized:  false
28:34.049 [off-pool-0] DEBUG triple.loader -- Initialize repository
28:34.049 [off-pool-0] ERROR triple.loader -- Initialise error [java.lang.IllegalStateException]: Sail has already been intialized. Ensure this Sail is being used via a Repository.
28:34.050 [off-pool-0] DEBUG triple.loader -- shutdown repository
```

IMHO it is cause either by  wrong propagation of the shutDown or shutdown from the old session is not finished yet. 

Full stacktrace is:
```
expected: nil
  actual: java.lang.IllegalStateException: Sail has already been intialized. Ensure this Sail is being used via a Repository.
 at org.eclipse.rdf4j.sail.helpers.AbstractSail.initialize (AbstractSail.java:174)
    org.eclipse.rdf4j.sail.helpers.SailWrapper.initialize (SailWrapper.java:99)
    org.eclipse.rdf4j.sail.lucene.LuceneSail.initialize (LuceneSail.java:321)
    org.eclipse.rdf4j.repository.sail.SailRepository.initializeInternal (SailRepository.java:151)
    org.eclipse.rdf4j.repository.base.AbstractRepository.initialize (AbstractRepository.java:31)
    triple.loader$eval2641$fn__2643.doInvoke (loader.clj:88)
    clojure.lang.RestFn.invoke (RestFn.java:521)
    clojure.lang.MultiFn.invoke (MultiFn.java:260)
    triple.loader$eval2635$fn__2637.doInvoke (loader.clj:77)
    clojure.lang.RestFn.invoke (RestFn.java:521)
    clojure.lang.MultiFn.invoke (MultiFn.java:260)
    triple.loader$eval2629$fn__2631.doInvoke (loader.clj:74)
    clojure.lang.RestFn.invoke (RestFn.java:425)
    clojure.lang.MultiFn.invoke (MultiFn.java:233)
    sparql$load_multidata.invokeStatic (sparql.clj:119)
    sparql$load_multidata.invoke (sparql.clj:108)
    triple.multiload_test$fn__2822.invokeStatic (multiload_test.clj:17)
    triple.multiload_test/fn (multiload_test.clj:15)
    clojure.test$test_var$fn__7983.invoke (test.clj:716)
    clojure.test$test_var.invokeStatic (test.clj:716)
    clojure.test$test_var.invoke (test.clj:707)
    boot.user$eval1804$fn__1805.doInvoke (boot.user6721134797804565458.clj:23)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:29)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:646)
    clojure.core$apply.invoke (core.clj:641)
    boot.core$construct_tasks.doInvoke (core.clj:911)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:646)
    clojure.core$apply.invoke (core.clj:641)
    boot.core$boot$fn__933.invoke (core.clj:949)
    clojure.core$binding_conveyor_fn$fn__4676.invoke (core.clj:1938)
    clojure.lang.AFn.call (AFn.java:18)
    java.util.concurrent.FutureTask.run (FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:617)
    java.lang.Thread.run (Thread.java:745)
```

### PR DESCRIPTION
**Pull request by jgrzebyta**
This PR addresses GitHub issue: #607 .

Briefly describe the changes proposed in this PR:

* After doing shutdown() user is not able to start new connection - it will throw exception
* Added more loggers

Signed-off-by:Jacek Grzebyta <grzebyta.dev@gmail.com>

### CODE REVIEW

File: core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java [hunk 1]
```diff
@@ -282,14 +282,20 @@ public SearchIndex getLuceneIndex() {
 	public NotifyingSailConnection getConnection()
 		throws SailException
 	{
-		return new LuceneSailConnection(super.getConnection(), luceneIndex, this);
+		if (!closed.get()) {
+			return new LuceneSailConnection(super.getConnection(), luceneIndex, this);
+		}
+		else {
+			throw new SailException("Connection cannot be taken from shutdown sail");
```
Comment (jeenbroekstra) [inline core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java]: I'd rephrase this message to "Sail is shut down or not initialized".

File: core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java [hunk 2]
```diff
@@ -282,14 +282,20 @@ public SearchIndex getLuceneIndex() {
...
+		}
 	}
 
 	@Override
 	public void shutDown()
 		throws SailException
 	{
-		if(closed.compareAndSet(false, true)) {
+		logger.debug("shutdown");
```
Comment (jeenbroekstra) [inline core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java]: We need to be a little careful with what kind of debug logging we put in place - it's a big framework and it can quickly get messy and overwhelming.  Perhaps rephrase to "LuceneSail shutDown".

File: core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/AbstractSail.java [hunk 3]
```diff
@@ -198,6 +199,7 @@ public void shutDown()
 		throws SailException
 	{
 		initializationLock.writeLock().lock();
+		logger.debug("shutdown ...");
```
Comment (jeenbroekstra) [inline core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/AbstractSail.java]: Can you remove this?

File: core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailRepository.java [hunk 4]
```diff
@@ -166,6 +166,7 @@ protected void shutDownInternal()
 		throws RepositoryException
 	{
 		try {
+			logger.debug("Sail: {}", sail);
```
Comment (jeenbroekstra) [inline core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailRepository.java]: Rephrase to "Shutting down Sail: {}" - btw does sail.toString() give a useful, readable message?

**Comment by jgrzebyta** [comment]
ok. Done

**Comment by jeenbroekstra** [comment]
Thanks Jacek, LGTM.

**Comment by jeenbroekstra** [comment]
Running hudson build on PR: https://hudson.eclipse.org/rdf4j/job/rdf4j-verify-pr/33/

File: core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/AbstractSail.java [hunk 5]
```diff
@@ -169,6 +169,7 @@ public void initialize()
 		throws SailException
 	{
 		initializationLock.writeLock().lock();
+		logger.trace("is initialized: {}", isInitialized());
```
Comment (ansell) [inline core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/AbstractSail.java]: This call needs to be inside of the try block.

File: core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java [hunk 6]
```diff
@@ -282,14 +282,20 @@ public SearchIndex getLuceneIndex() {
...
+			throw new SailException("Sail is shut down or not initialized");
+		}
 	}
 
 	@Override
 	public void shutDown()
 		throws SailException
 	{
-		if(closed.compareAndSet(false, true)) {
+		logger.debug("LuceneSail shutdown");
```
Comment (ansell) [inline core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java]: This call should be inside of the if statement so that it only ever occurs once.

**Comment by jgrzebyta** [comment]
@ansell  Both changes are ok to me. For now how can I get your changes back into my local repo? I would like to merge with master now.
